src/lambda/lambda.lisp

```diff
@@ -1 +1,125 @@
+    ((right lty term)    (coprod lty (type-of1 ctx term)))
+    ((case-on on ltm)       (if (typep (type-of1 ctx on) 'coprod)
+                             (type-of1 (cons  (mcar (type-of1 ctx on)) ctx) ltm)
+                             (error "type of on not of coproduct type")))
+    ((pair ltm rtm)      (prod (type-of1 ctx ltm) (type-of1 ctx rtm)))
+    ((fst term)          (if (typep (type-of1 ctx term) 'prod)
+                             (mcar (type-of1 ctx term))
+                             (error "type of term not of product type")))
+    ((snd term)          (if (typep (type-of1 ctx term) 'prod)
+                             (mcadr (type-of1 ctx term))
+                             (error "type of term not of product type")))
+    ((lamb tdom term)    (fun-type tdom
+                                   (type-of1 (cons tdom ctx) term)))
+    ((app fun)           (hom-cod ctx fun))
+    ((index pos)         (index-check pos ctx))))
+
+;; Changes exp-aux to so-hom-obj
+
+(defun type-of2 (t1)
+    (cond ((typep t1 'prod)     (prod (type-of2 (mcar t1))
+                                      (type-of2 (mcadr t1))))
+          ((typep t1 'coprod)   (coprod (type-of2 (mcar t1))
+                                        (type-of2 (mcadr t1))))
+          ((typep t1 'fun-type) (so-hom-obj (type-of2 (mcar t1))
+                                            (type-of2 (mcadr t1))))
+          (t                    t1)))
+
+(defun inf-pass (context term)
+  (type-of2 (type-of1 context term)))
+
+;; Annotated type-inferencer. Types are accessible through ttype function
+
+;;Problem with equalp here, does not recognize same types, e.g. in pair case after projecting
+
+(defmethod well-defp (ctx (tterm <stlc>))
+  (match-of stlc tterm
+    ((absurd term)       (and (well-defp ctx term)
+                                   (geb.mixins:obj-equalp (type-of1 ctx term) so0)))
+    (unit                     t)
+    ((left term)          (and (well-defp ctx term)
+                                   (geb.mixins:obj-equalp (mcar
+                                            (type-of1 ctx tterm))
+                                           (type-of1 ctx term))))
+    ((right term)         (and (well-defp ctx term)
+                                   (geb.mixins:obj-equalp (mcadr
+                                            (type-of1 ctx tterm))
+                                           (type-of1 ctx term))))
+    ((case-on on ltm rtm)  (and (well-defp (cons  (mcar (type-of1 ctx on)) ctx)
+                                           ltm)
+                                (well-defp (cons  (mcadr (type-of1 ctx on)) ctx)
+                                           rtm)
+                                (well-defp ctx on)))
+    ((pair ltm rtm)        (and (well-defp ctx ltm)
+                                (well-defp ctx rtm)
+                                (geb.mixins:obj-equalp (mcar (type-of1 ctx tterm))
+                                        (type-of1 ctx ltm))
```

Comment:
Some of the formatting here is suspect, because arguments need to align, it's not worth detracting the PR, but yeah